### PR TITLE
Introduce paste from libre office plugin

### PR DESCRIPTION
**Pull request by msamsel**
## What is the purpose of this pull request?

New feature

## Does your PR contain necessary tests?

All patches which change the editor code must include tests. You can always read more
on [PR testing](https://ckeditor.com/docs/ckeditor4/latest/guide/dev_contributing_code.html#tests),
[how to set the testing environment](https://ckeditor.com/docs/ckeditor4/latest/guide/dev_tests.html) and
[how to create tests](https://ckeditor.com/docs/ckeditor4/latest/guide/dev_tests.html#creating-your-own-test)
in the official CKEditor documentation.

### This PR contains

- [x] Unit tests
- [x] Manual tests

## What is the proposed changelog entry for this pull request?

```
Added support for pasting rich content from Libre Office with the [Paste from Libre Office](https://ckeditor.com/cke4/addon/pastefromlibreoffice) plugin.
```

## What changes did you make?

Requires: #3593

Implement part of the content fixing mechanism inside the common filter. Thi result with fixing some of the wrong expected output for IE11 & Safari on PFW.
The plugin doesn't change pftools mechanism, however, it bypasses it in some places.

### CODE REVIEW

File: plugins/pastefromlibreoffice/filter/default.js [hunk 1]
```diff
@@ -0,0 +1,314 @@
...
+
+				elementNames: [
+					[ /^head$/i, '' ],
+					[ /^meta$/i, '' ],
+					[ /^strike$/i, 's' ]
+				],
+
+				elements: {
+					'!doctype': function( el ) {
+						el.replaceWithChildren();
```
[the remark this answers is not in the document]
Comment (msamsel) [reply plugins/pastefromlibreoffice/filter/default.js]: So we have bug in html parser then:
https://codepen.io/msamsel/pen/OJJdWLv?editors=0011

[the remark this answers is not in the document]
Comment (Comandeer) [reply plugins/pastefromlibreoffice/filter/default.js]: @msamsel, can you report it?

[the remark this answers is not in the document]
Comment (msamsel) [reply plugins/pastefromlibreoffice/filter/default.js]: Extracted to #3664

File: plugins/pastefromlibreoffice/filter/default.js [hunk 2]
```diff
@@ -0,0 +1,314 @@
...
+
+					'border': remove
+				}
+			};
+		}
+	};
+
+	function replaceEmptyElementWithChildren( element ) {
+		if ( !CKEDITOR.tools.object.entries( element.attributes ).length ) {
+			element.replaceWithChildren();
```
[the remark this answers is not in the document]
Comment (msamsel) [reply plugins/pastefromlibreoffice/filter/default.js]: For this case, both methods work in the same way. My approach, however, doesn't require to return `false` from the filter, so it seems to be simpler. That's why I prefer the current solution. If you find some cases where this function behaves not intuitive, then I change it.

File: plugins/pastefromword/plugin.js [hunk 3]
```diff
@@ -25,7 +25,7 @@
 				configInlineImages = editor.config.pasteFromWord_inlineImages === undefined ? true : editor.config.pasteFromWord_inlineImages,
 				defaultFilters = [
 					CKEDITOR.getUrl( pastetoolsPath + 'filter/common.js' ),
-					CKEDITOR.getUrl(  path + 'filter/default.js' )
+					CKEDITOR.getUrl( path + 'filter/default.js' )
```
[the remark this answers is not in the document]
Comment (msamsel) [reply plugins/pastefromword/plugin.js]: Unfortunately not.

I tried to persist PFW and its logic as close to the source as I could. As the main scope of the feature is to deliver Paste From Libre Office not to modify/unify/enhance/etc. PFW. That's why I don't modify the way how image processing is working for PFW. However, I extract the image processing logic from Libre Office to a general filter, to merge those two much easier one day. I think that better would be to cover it in a separate issue.

Another thing is that the regexps, which are responsible for image extraction, have a difference between Word and Libre Office. It looks like, the RTF content is not created exactly the same way for both editors.

Having those things on my mind I decided to not include merging logic responsible for the image processing in Word and LIbre office together with this PR.

File: plugins/pastetools/filter/common.js [hunk 4]
```diff
@@ -797,6 +826,100 @@
 		}
 	}
 
+	function removeSuperflousStyles( element ) {
+		var resetStyles = [
+				'background-color:transparent',
+				'background:transparent',
+				'background-color:none',
+				'background:none',
+				'text-align:start',
```
[the remark this answers is not in the document]
Comment (msamsel) [reply plugins/pastetools/filter/common.js]: Probably some remaining leftover from the early stage of development.

File: tests/plugins/pastetools/filter/common.js [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+	bender.test( {
+		'tests common filter should remove apple spaces': function() {
+			assertCommonFilter(
+				this.editor,
+				'<p>foo<span class="Apple-converted-space">\u00A0</span>bar</p>',
+				'<p>foo bar</p>'
+			);
+		},
+
+		'test common filter should remove superflous styles': function() {
```
[the remark this answers is not in the document]
Comment (msamsel) [reply tests/plugins/pastetools/filter/common.js]: I missed that during batch commit :( So I add it separately to not spend too much time on git history fixing.

**Comment by msamsel** [comment]
> There are some failing tests:

It seems to pass on my env. Please check if those still fails after rebase to current major and if so, then provide more details on what browser and which version you are using.

**Comment by jacekbogdanski** [comment]
@msamsel I see that this PR introduces some changes to image handling. Could you confirm if we should wait with https://github.com/ckeditor/ckeditor4/issues/2800 until this PR gets merged?

**Comment by msamsel** [comment]
@jacekbogdanski this PR is not related to #2800.
PFLO doesn't implement any changes to image processing in PFW. From my perspective linked issue can be handled separately, as long, as it **only** affects processing images. In case that proposed solution would require changes in the PFW filter, then would be better to wait for merge.

**Comment by msamsel** [comment]
I've rebased the pr to current `major` and resolve existing conflicts.

**Comment by msamsel** [comment]
Issue related to `float` style added to images is extracted to: #3780

**Comment by msamsel** [comment]
> On Safari image is not inserted at all. 

Safari doesn't provide the image, that's why is not possible to obtain it.
There is no `text/rtf` clipboard in the paste event and there is no `src` reference in `text/html` clipboard.
Screenshot 2020-01-15 at 12 42 18

**Comment by msamsel** [comment]
The case related to missing images on IE11 is extracted to #3781. The image cannot be embedded in content as there is no available data which allows on that. This is why most suitable fix for this case is removing such elements, however, this change will affect the PFW behaviour.

**Comment by msamsel** [comment]
About IE8 I remain it as it is. PFLO is a new feature and those has limited compatibility with obsolete browsers. It doesn't look that breaks something on this browser.

File: plugins/pastefromword/plugin.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (msamsel) [reply plugins/pastefromword/plugin.js]: #3782 3782

**Comment by Comandeer** [comment]
> Safari doesn't provide the image, that's why it is not possible to obtain the image.

I've reported it: https://bugs.webkit.org/show_bug.cgi?id=206348

**Comment by msamsel** [comment]
> It seems that SVG file is inside RTF as `pict` element. Maybe we would be able to add the support easily 🤔

Unfortunately, the format of this is image is [`wmetafile8`](https://en.wikipedia.org/wiki/Rich_Text_Format#Pictures). It would require to write some dedicated parser/converter which would allow embedding such image in the document. It is not a clear SVG image.
When there are used png or gif images then also for each image has it's `wmf` representation in rtf clipboard. AFAIR I tried to analyze this format to have possibility to embed shapes drawn in MS Word unfortunately without luck.

**Comment by LiamDawe** [comment]
Does this need a new toolbar button or is it added into other paste features?

**Comment by f1ames** [comment]
@LiamDawe - it's added by default in Standard and Full presets and doesn't use additional paste button. So all content pasted from Libre Office with `Ctrl`/`Cmd`+`C`, patse button or context menu will go through Libre Office filter preserving formatting.

**Pull request merged**